### ISSUE TITLE
'NoneType' object has no attribute 'replace'

### ISSUE THREAD

**Issue by codingkidprodigy** [iMakedonsky/drf-autodocs#24]
Hello,
I have been trying to use DRF autodocs with my Django Rest Framework REST API, but every time I set up my project to use it I get the following error:
'NoneType' object has no attribute 'replace'.
The URL I am using is localhost:8000/authenticate/docs
I even tried: localhost:8000/docs/ but it didn't work.
Here is the contents of my views.py file.
Thanks for your help!
```

from .serializers import (
    LoginSerializer,
    RegistrationSerializer,
    UserSerializer
)
from .renderers import UserJSONRenderer
from rest_framework import status
from rest_framework.generics import RetrieveUpdateAPIView
from rest_framework.permissions import AllowAny, IsAuthenticated
from rest_framework.response import Response
from rest_framework.views import APIView
# Create your views here.


class RegistrationAPIView(APIView):
    permission_classes = (AllowAny,)
    renderer_classes = (UserJSONRenderer,)
    serializer_class = RegistrationSerializer

    def post(self, request):
        user = request.data.get('user', {})

        serializer = self.serializer_class(data=user)
        serializer.is_valid(raise_exception=True)
        serializer.save()

        return Response(serializer.data, status=status.HTTP_201_CREATED)


class LoginAPIView(APIView):
    permission_classes = (AllowAny,)
    renderer_classes = (UserJSONRenderer,)
    serializer_class = LoginSerializer

    def post(self, request):
        user = request.data.get('user', {})
        serializer = self.serializer_class(data=user)
        serializer.is_valid(raise_exception=True)

        return Response(serializer.data, status=status.HTTP_200_OK)


class UserRetrieveUpdateAPIView(RetrieveUpdateAPIView):
    permission_classes = (IsAuthenticated,)
    renderer_classes = (UserJSONRenderer,)
    serializer_class = UserSerializer

    def retrieve(self, request, *arg, **kwargs):
        serializer = self.serializer_class(request.user)

        return Response(serializer.data, status=status.HTTP_200_OK)

    def update(self, request, *arg, **kwargs):
        serializer_data = request.data.get('user', {})

        serializer = self.serializer_class(
            request.user, data=serializer_data, partial=True
        )
        serializer.is_valid(raise_exception=True)
        serializer.save()

        return Response(serializer.data, status=status.HTTP_200_OK)

```

**Comment by ZandekA** [comment]
Hello, @codingkidprodigy can you show your urls.py files?

**Comment by codingkidprodigy** [comment]
Recently, I switched over to a new computer and I set up everything again. But, now I get this error:
ModuleNotFoundError: No module named 'django.core.urlresolvers'

I have a MacBook Pro and I have Django Version 2.0 and Python Version 3.6.4

**Comment by ZandekA** [comment]
@codingkidprodigy First link in google: https://stackoverflow.com/a/43139407

**Comment by codingkidprodigy** [comment]
I have followed the instructions there, but now I have this error:
from django.core.urlresolvers import RegexURLResolver, RegexURLPattern
ModuleNotFoundError: No module named 'django.core.urlresolvers'

**Comment by codingkidprodigy** [comment]
So, after spending some time working on this problem, I found out something. Here is my app url:
``
path('', include('drf_autodocs.urls'), name="docs"),
path('user/', UserRetrieveUpdateAPIView.as_view(), name="user_view"),
path('users/register/', RegistrationAPIView.as_view(), name="register"),
path('users/login/', LoginAPIView.as_view(), name="login"),
``

My root url conf is:
``
urlpatterns = [
    path('authenticate/', include('authentication.urls')),
]
``

So, I get the error above when I keep the `path('', include('drf_autodocs.urls'), name="docs"`
If I remove that line then I do not get the error above. I have a feeling that DRF autodocs has to be updated to work with Django 2.0. Is there any fix for this. I also think that DRF Autodocs uses the REGEX and Django 2.0 does not. So .... I don't know what to do. Thanks for any help in advance

**Comment by ZandekA** [comment]
@codingkidprodigy When you **includes** urls from some app, you should assign **namespace** for this group of urls:
`path('', include('drf_autodocs.urls'), namespace="docs")`
When you **defines** an endpoint url you should assing **name** for this route: `path('user/', UserRetrieveUpdateAPIView.as_view(), name="user_view")`

**Comment by codingkidprodigy** [comment]
I have kept the namespace for the docs but the same error is coming:

from django.core.urlresolvers import RegexURLResolver, RegexURLPattern
ModuleNotFoundError: No module named 'django.core.urlresolvers'

**Comment by ZandekA** [comment]
@codingkidprodigy Can you copy full error log?

**Comment by codingkidprodigy** [comment]
Here is the full error log:
Performing system checks...

Unhandled exception in thread started by <function check_errors.<locals>.wrapper at 0x10d5c3a60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/utils/autoreload.py", line 225, in wrapper
    fn(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/commands/runserver.py", line 120, in inner_run
    self.check(display_num_errors=True)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 364, in check
    include_deployment_checks=include_deployment_checks,
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 351, in _run_checks
    return checks.run_checks(**kwargs)
  File "/usr/local/lib/python3.6/site-packages/django/core/checks/registry.py", line 73, in run_checks
    new_errors = check(app_configs=app_configs)
  File "/usr/local/lib/python3.6/site-packages/django/core/checks/urls.py", line 13, in check_url_config
    return check_resolver(resolver)
  File "/usr/local/lib/python3.6/site-packages/django/core/checks/urls.py", line 23, in check_resolver
    return check_method()
  File "/usr/local/lib/python3.6/site-packages/django/urls/resolvers.py", line 397, in check
    for pattern in self.url_patterns:
  File "/usr/local/lib/python3.6/site-packages/django/utils/functional.py", line 36, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/usr/local/lib/python3.6/site-packages/django/urls/resolvers.py", line 536, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/usr/local/lib/python3.6/site-packages/django/utils/functional.py", line 36, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/usr/local/lib/python3.6/site-packages/django/urls/resolvers.py", line 529, in urlconf_module
    return import_module(self.urlconf_name)
  File "/usr/local/Cellar/python3/3.6.4/Frameworks/Python.framework/Versions/3.6/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/abhinavvemulapalli/Documents/PavaC/back-end/backend/urls.py", line 19, in <module>
    path('authenticate/', include('authentication.urls')),
  File "/usr/local/lib/python3.6/site-packages/django/urls/conf.py", line 34, in include
    urlconf_module = import_module(urlconf_module)
  File "/usr/local/Cellar/python3/3.6.4/Frameworks/Python.framework/Versions/3.6/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/abhinavvemulapalli/Documents/PavaC/back-end/authentication/urls.py", line 10, in <module>
    path('docs/', include('drf_autodocs.urls'), namespace="docs"),
  File "/usr/local/lib/python3.6/site-packages/django/urls/conf.py", line 34, in include
    urlconf_module = import_module(urlconf_module)
  File "/usr/local/Cellar/python3/3.6.4/Frameworks/Python.framework/Versions/3.6/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/lib/python3.6/site-packages/drf_autodocs/urls.py", line 2, in <module>
    from .views import TreeView
  File "/usr/local/lib/python3.6/site-packages/drf_autodocs/views.py", line 2, in <module>
    from .parser import TreeAPIParser
  File "/usr/local/lib/python3.6/site-packages/drf_autodocs/parser.py", line 4, in <module>
    from django.core.urlresolvers import RegexURLResolver, RegexURLPattern
ModuleNotFoundError: No module named 'django.core.urlresolvers'

**Comment by ZandekA** [comment]
So, don't use Windows, use Linux or OSX. Be true developer :D. You have problems with imports, because drf_autodocs can't find Django.

**Comment by codingkidprodigy** [comment]
As a matter of fact, I am using OSX. I recently switched over to Mac and the problem still persists

**Comment by ZandekA** [comment]
Sorry, it was a sarcasm. So, as I told it looks like a problem with import, because error occurs inside exception.

**Comment by codingkidprodigy** [comment]
No problem, how will I be able to fix the import problem. I am a little bit of a newbie when it comes to Django

**Comment by codingkidprodigy** [comment]
How will I be able to fix the import problem? I don't really know why drf_autodocs can't find Django. I know everything is setup perfectly. I am using Mac so ...... I think it might be that I have to use python3 instead of python so. I don't know

**Comment by codingkidprodigy** [comment]
I actually think I found the problem but I do not know how to fix it. It will require knowledge about how DRF_Autodocs was written. So onto what the problem is:
In parsers.py where my package was downloaded. RegexURLResolver and RegexURLPattern are being imported from django.core.urlresolvers. I changed it to django.urls and it still doesn't work because those 2 imports weren't found. So, the author of DRF_Autodocs has to change it so that the Tree API View is not dependent on RegexURLResolver and RegexURLPattern. If there is any way to fix it please do tell me.

**Comment by codingkidprodigy** [comment]
I have solved the problem. It required altering the parser.py file where my package was located. Which means I will have to send a merge request with the changes. But here are the changes and what went wrong. In parsers.py I had to import 2 different things from django.urls.resolvers:
RegexURLResolver was replaced with RegexPattern
and RegexURLPattern was replaced with RoutePattern. After that it should work. I will try to send a merge request.
